components/datetime/src/options/components.rs

```diff
@@ -0,0 +1,209 @@
+use crate::fields::{self, Field, FieldLength, FieldSymbol, FieldType};
+use std::fmt;
+
+#[derive(Debug)]
+pub struct Bag {
+    pub era: Text,
+    pub year: Numeric,
+    pub month: Month,
+    pub day: Numeric,
+    pub weekday: Text,
+
+    pub hour: Numeric,
+    pub minute: Numeric,
+    pub second: Numeric,
+    pub hour_cycle: HourCycle,
```

Comment:
In ECMA402, we have `hour_cycle` with:

```
pub enum HourCycle {
    H24,
    H23,
    H12,
    H11,
}
```

Are you suggesting just renaming, or are the variants then different as well? Is it your intention to diverge from ECMA402?

Reply:
Maybe I'm not understanding the purpose of this struct. It looks to me like the main public options bag for skeletons. In that case, ECMA 402 has a day period option only. It has an hour cycle option only in the context of resolving the locale, which itself carries an hour cycle.

Reply:
From: https://tc39.es/ecma402/#datetimeformat-objects

> 13.1.1 InitializeDateTimeFormat
>> 13. Let hourCycle be ? GetOption(options, "hourCycle", "string", « "h11", "h12", "h23", "h24" », undefined).


Am I missing something?

Reply:
That setting is in the same category as calendar and numbering system. It's a user preference, not a formatting style setting

Reply:
in JS developer can write:

```js
let dtf = new Intl.DateTimeFormat(langid, {
  dateStyle: "long",
  timeStyle: "medium",
  hourCycle: "h23"
});
dtf.format(date);
```

In this PR, user can write:

```rust
let dtf = DateTimeFormat::try_new(langid, &provider, &options::DateTimeFormatOptions::Style(options::Style::Bag {
  date_style: options::style::Date::Long,
  time_style: options::style::Time::Medium,
  hour_cycle: options::components::HourCycle::H23
})).unwrap();
dtf.format_to_string(date);
```

I was aiming for keeping the API close. What would you prefer me to do?

Reply:
JavaScript developers can also write

```javascript
let dtf = new Intl.DateTimeFormat("fa", {
  dateStyle: "long",
  timeStyle: "medium",
  dayPeriod: "long",
});
dtf.format(new Date());
```

but you don't have a dayPeriod option in your bag.

hourCycle is part of the locale: `en-US-u-hc-h23` for 23-hour time in en-US, and `fr-FR-u-hc-12` for 12-hour time in fr-FR.  We should think more about how we want to handle the class of options in ECMA-402 that are not really options for the formatter, but are actually overrides for the locale.

Reply:
Would you like me to just remove `hour_cycle` from this PR for now?

Reply:
Let's talk about it on Thursday

Reply:
I think that we should have options for everything, even if it is possible to represent as part of the locale.
There should be no difference between "overrides" or not.

It not not impossible that the locale extension `-u-` adds some extra option, which we already have in the bag.
It was not override when we designed the API, but now it is.
So we end up with options that are overrides, and options that are not.

Reply:
But... I agree that `hourCycle` does no belong together with `dateStyle` / `timeStyle`.
This is an area where I am not sure I like what ECMAScript did.

In ICU the difference was clear. You could construct a formatter with:
* date / time styles (short / medium / long / full)
* patterns
* skeletons
But not a mixture.

ECMAScript seems to allows mixtures.
You can do:
```
var df = new Intl.DateTimeFormat(locale, { dateStyle:'long' });
```
and
```
var df = new Intl.DateTimeFormat(locale, { month: 'numeric' });
```

There is nothing to tell you if you can or can't do
```
var df = new Intl.DateTimeFormat(locale, { dateStyle:'long', month: 'numeric' });
```

Firefox throws: "Uncaught TypeError: can't set option month when dateStyle is used"
Chrome throws too, with a more obscure message:
"VM439:1 Uncaught TypeError: Invalid option : dateStyle
    at new DateTimeFormat (<anonymous>)
    at <anonymous>:1:10"

But why runtime?
My preference is to always make it clear from the API signatures...

Even "funnier", `hourCyle` can be used with both styles, and bag-of-options.

---

I think the distinction "user preference" vs "style" is somewhat artificial, and does no really help.

Both Windows and MacOS allows the user the explicitly edit the date / time patterns.
They did that "since forever", using publicly accessible UI, not some hack.
So "3 Mar 2020" vs "March 3, 2020" is a user preference, if we look at the 2 most popular end-user OSes.

Reply:
Added `Preferences`